### DISCUSSION TITLE
How to transfer lima vm instance to another machine?

### DISCUSSION THREAD

**Discussion by adonishong** [lima-vm/lima#1639] [Q&A] [answered]
As described in title, how to transfer lima vm instance to another machine?

I have search issues, also went through lima Q&A, get no result ...

Let's say I want to transfer "default" instance to another machine, I have tried copy full directory of ~/.lima/default to target machine. Both machine have same version of lima, but no luck with this way. 

Here is the output I get:
INFO[0000] Using the existing instance "default_ext"    
INFO[0000] Attempting to download the nerdctl archive    arch=aarch64 digest="sha256:589dabd962d936b29fd377dcddbb49c07d1c4c27dd4b402bc4b6b20287fe9c37" location="https://github.com/containerd/nerdctl/releases/download/v1.4.0/nerdctl-full-1.4.0-linux-arm64.tar.gz"
INFO[0000] Using cache "/Users/adonishong/Library/Caches/lima/download/by-url-sha256/8704652b907393aea54c8d98b48efa0fe9bef4d120fbbf8799bb5264631da634/data" 
INFO[0000] [hostagent] Starting QEMU (hint: to watch the boot progress, see "/Users/adonishong/.lima/default_ext/serial.log") 
INFO[0000] SSH Local Port: 56729                        
INFO[0000] [hostagent] Waiting for the essential requirement 1 of 5: "ssh" 
INFO[0000] [hostagent] Driver stopped due to error: "exit status 1" 
INFO[0000] [hostagent] Shutting down the host agent     
WARN[0000] [hostagent] failed to exit SSH master        
INFO[0000] [hostagent] Shutting down QEMU with ACPI     
WARN[0000] [hostagent] failed to open the QMP socket "/Users/adonishong/.lima/default_ext/qmp.sock", forcibly killing QEMU 
INFO[0000] [hostagent] QEMU has already exited          
FATA[0000] exiting, status={Running:false Degraded:false Exiting:true Errors:[] SSHLocalPort:0} (hint: see "/Users/adonishong/.lima/default_ext/ha.stderr.log")

###########################################################
### **Let me make a quick conclusion till 06/26/23.**
My current conclusion is, transfer lima vm instance to another machine need three operation
1. This is the most important part, username and home directory in both machine (source and target) should be strictly same. If there is any difference in this part, transfer will be failed ...
2. Copy the original instance directory to target machine, let's say we want to transfer "default" instance, then, you need to copy ~/.lima/default directory
3. Copy the ~/.lima/_config directory to target machine

Thanks to the responses from @jandubois @AkihiroSuda, in case of mislead the other reader, I will not mark your response as answer, but still appreciate for your reply.

I am still looking for the other way to transfer instance without username and home path keep same requirement. It is welcome to send me message if there is any other update.

**Comment by jandubois** [comment]
You'll need to copy at least your `~/.lima/_config/user` and `~/.lima/_config/user.pub` keys as well, but preferably the whole `_config` directory.

**Reply by adonishong** [reply]
I have done that also, in fact, I have copied the whole ~/.lima directory with all subdirectory to target machine. With all this done, still not work, same error as first thread.

**Reply by AkihiroSuda** [reply]
`/Users/adonishong/.lima/default_ext/ha.*.log` may contain more information

**Reply by adonishong** [reply]
Thanks for your tips ^_^, from ha.stderr.log, I found out that this instance will look for disk data with absolute path from source machine, and in target machine the user name is not same, so I create a link to fix this issue. Now, I get new issue, which is different to the first situation. Here is the ha.stderr.log. I have noticed the error "{"level":"debug","msg":"stdout=\"\", stderr=\"adonishong@127.0.0.1: Permission denied (publickey).\\r\\n\", err=failed to execute script \"ssh\": stdout=\"\", stderr=\"adonishong@127.0.0.1: Permission denied (publickey).\\r\\n\": exit status 255","time":"2023-06-21T13:08:54+08:00"}", but I have no idea how to fix this issue ...

(base) adonishongs-MacBook-Pro:.lima adonishong$ cat /Users/adonishong/.lima/default/ha.stderr.log 
{"level":"debug","msg":"Creating iso file /Users/adonishong/.lima/default/cidata.iso","time":"2023-06-21T13:08:22+08:00"}
{"level":"debug","msg":"Using /var/folders/68/7mwry6ks2l15w3gp88096bnc0000gn/T/diskfs_iso3934483670 as workspace","time":"2023-06-21T13:08:22+08:00"}
{"level":"debug","msg":"Failed to detect CPU features. Assuming that AES acceleration is available on this Apple silicon.","time":"2023-06-21T13:08:23+08:00"}
{"level":"debug","msg":"OpenSSH version 9.0.1 detected","time":"2023-06-21T13:08:23+08:00"}
{"level":"debug","msg":"AES accelerator seems available, prioritizing aes128-gcm@openssh.com and aes256-gcm@openssh.com","time":"2023-06-21T13:08:23+08:00"}
{"level":"debug","msg":"Start udp server listening on: 127.0.0.1:60751","time":"2023-06-21T13:08:23+08:00"}
{"level":"debug","msg":"Start tcp server listening on: 127.0.0.1:58300","time":"2023-06-21T13:08:23+08:00"}
{"level":"debug","msg":"QEMU version 8.0.2 detected","time":"2023-06-21T13:08:23+08:00"}
{"level":"debug","msg":"firmware candidates = [/Users/adonishong/.local/share/qemu/edk2-aarch64-code.fd /opt/homebrew/share/qemu/edk2-aarch64-code.fd /usr/share/AAVMF/AAVMF_CODE.fd /usr/share/qemu-efi-aarch64/QEMU_EFI.fd]","time":"2023-06-21T13:08:23+08:00"}
{"level":"info","msg":"Starting QEMU (hint: to watch the boot progress, see \"/Users/adonishong/.lima/default/serial.log\")","time":"2023-06-21T13:08:23+08:00"}
{"level":"debug","msg":"qCmd.Args: [/opt/homebrew/bin/qemu-system-aarch64 -m 8192 -cpu host -machine virt,accel=hvf -smp 8,sockets=1,cores=8,threads=1 -drive if=pflash,format=raw,readonly=on,file=/opt/homebrew/share/qemu/edk2-aarch64-code.fd -boot order=c,splash-time=0,menu=on -drive file=/Users/adonishong/.lima/default/diffdisk,if=virtio,discard=on -drive id=cdrom0,if=none,format=raw,readonly=on,file=/Users/adonishong/.lima/default/cidata.iso -device virtio-scsi-pci,id=scsi0 -device scsi-cd,bus=scsi0.0,drive=cdrom0 -netdev user,id=net0,net=192.168.5.0/24,dhcpstart=192.168.5.15,hostfwd=tcp:127.0.0.1:60022-:22 -device virtio-net-pci,netdev=net0,mac=52:55:55:14:75:52 -device virtio-rng-pci -display none -vga none -device ramfb -device qemu-xhci,id=usb-bus -device usb-kbd,bus=usb-bus.0 -device usb-mouse,bus=usb-bus.0 -parallel none -chardev socket,id=char-serial,path=/Users/adonishong/.lima/default/serial.sock,server=on,wait=off,logfile=/Users/adonishong/.lima/default/serial.log -serial chardev:char-serial -chardev socket,id=char-qmp,path=/Users/adonishong/.lima/default/qmp.sock,server=on,wait=off -qmp chardev:char-qmp -name lima-default -pidfile /Users/adonishong/.lima/default/qemu.pid]","time":"2023-06-21T13:08:23+08:00"}
{"level":"info","msg":"Waiting for the essential requirement 1 of 5: \"ssh\"","time":"2023-06-21T13:08:23+08:00"}
{"level":"debug","msg":"executing script \"ssh\"","time":"2023-06-21T13:08:23+08:00"}
{"level":"debug","msg":"executing ssh for script \"ssh\": /usr/bin/ssh [ssh -F /dev/null -o IdentityFile=\"/Users/adonishong/.lima/_config/user\" -o IdentityFile=\"/Users/adonishong/.ssh/id_rsa\" -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o NoHostAuthenticationForLocalhost=yes -o GSSAPIAuthentication=no -o PreferredAuthentications=publickey -o Compression=no -o BatchMode=yes -o IdentitiesOnly=yes -o Ciphers=\"^aes128-gcm@openssh.com,aes256-gcm@openssh.com\" -o User=adonishong -o ControlMaster=auto -o ControlPath=\"/Users/adonishong/.lima/default/ssh.sock\" -o ControlPersist=5m -p 60022 127.0.0.1 -- /bin/bash]","time":"2023-06-21T13:08:23+08:00"}
{"level":"debug","msg":"stdout=\"\", stderr=\"adonishong@127.0.0.1: Permission denied (publickey).\\r\\n\", err=failed to execute script \"ssh\": stdout=\"\", stderr=\"adonishong@127.0.0.1: Permission denied (publickey).\\r\\n\": exit status 255","time":"2023-06-21T13:08:43+08:00"}
{"level":"info","msg":"Waiting for the essential requirement 1 of 5: \"ssh\"","time":"2023-06-21T13:08:53+08:00"}
{"level":"debug","msg":"executing script \"ssh\"","time":"2023-06-21T13:08:53+08:00"}
{"level":"debug","msg":"executing ssh for script \"ssh\": /usr/bin/ssh [ssh -F /dev/null -o IdentityFile=\"/Users/adonishong/.lima/_config/user\" -o IdentityFile=\"/Users/adonishong/.ssh/id_rsa\" -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o NoHostAuthenticationForLocalhost=yes -o GSSAPIAuthentication=no -o PreferredAuthentications=publickey -o Compression=no -o BatchMode=yes -o IdentitiesOnly=yes -o Ciphers=\"^aes128-gcm@openssh.com,aes256-gcm@openssh.com\" -o User=adonishong -o ControlMaster=auto -o ControlPath=\"/Users/adonishong/.lima/default/ssh.sock\" -o ControlPersist=5m -p 60022 127.0.0.1 -- /bin/bash]","time":"2023-06-21T13:08:53+08:00"}
{"level":"debug","msg":"stdout=\"\", stderr=\"adonishong@127.0.0.1: Permission denied (publickey).\\r\\n\", err=failed to execute script \"ssh\": stdout=\"\", stderr=\"adonishong@127.0.0.1: Permission denied (publickey).\\r\\n\": exit status 255","time":"2023-06-21T13:08:54+08:00"}

**Reply by AkihiroSuda** [reply]
You have to import `~/.lima/_config` from the old machine.

**Reply by adonishong** [reply]
Thank you for your reply ^_^ 

I copied the while "~/.lima" directory from source machine from very beginning ...

In fact, I have tried these combination
~/.lima/_config from source machine, ~/.ssh from source machine
~/.lima/_config from source machine, ~/.ssh from target machine
~/.lima/_config generated in target machine, ~/.ssh in target machine

I also insert ~/.ssh/id_rsa.pub from target machine into ~/.ssh/authorized_keys inside lima vm instance

None of these work lead to success ...

Due to current situation is vm is up, only ssh not work, is there anyway I could log into vm without ssh support? I also tried set display to "cocoa" in lima.yaml. But I could only get a window, there is no condole with it, only have some information start with "EFI stub: ...", last line is "EFI stub: Existing boot services..."

BTW, the OS inside this vm is Ubuntu 23.04

**Reply by adonishong** [reply]
Due to the error is "Permission denied (publickey)", I guess the issue come from username. The user name I use in source machine is "adonis", and the user name in target machine is "adonishong". I create a new user in target machine with origina user name "adonishong", and start lima vm with this user name ... No issue now, everything works ...

My new issue is, when I transfer a VM to a new machine with new user name, what shall I do ... seems like some info related to environment mix into vm instance ...

**Comment by adonishong** [comment, marked as answer]
My current conclusion is, transfer lima vm instance to another machine need three operation

This is the most important part, username and home directory in both machine (source and target) should be strictly same. If there is any difference in this part, transfer will be failed ...
Copy the original instance directory to target machine, let's say we want to transfer "default" instance, then, you need to copy ~/.lima/default directory
Copy the ~/.lima/_config directory to target machine
Thanks to the responses from @jandubois @AkihiroSuda, in case of mislead the other reader, I will not mark your response as answer, but still appreciate for your reply.

I am still looking for the other way to transfer instance without username and home path keep same requirement. It is welcome to send me message if there is any other update.

**Reply by cameronhudson8** [reply]
For those who have container volumes you want to preserve, you can do so like this:

1. Copy the data from the old guest VM to the host.
    1. Determine the mountpoint of the volume within the old guest.
        ```
        limactl shell <old-vm-name> nerdctl volume inspect <volume-name>
        ```
    1. Copy the data from the old guest to the host.
        ```
        limactl copy -r <old-vm-name>:<old-mountpoint> ~/container-volume-data
        ```
1. Copy the data from the host to the new guest VM.
    1. Create an empty volume with the same name as the old one.
        ```
        limactl shell <new-vm-name> nerdctl volume create <volume-name>
        ```
    1. Determine the mountpoint of the volume within the new guest.
        ```
        limactl shell <new-vm-name> nerdctl volume inspect <volume-name>
        ```
    1. Copy the data from the host into the new guest.
        ```
        limactl copy -r ~/container-volume-data <new-vm-name>:<new-mountpoint>
        ```
1. Start the container on the new guest VM and confirm that all is well.
1. Clean up.
    ```
    rm -rf ~/container-volume-data
    ```

If you have both VMs running on the same host, you also have the option of performing a `limactl copy` from the old guest directly to the new guest.

**Comment by be-a-bee** [comment]
As this post is quite old. Just wondering if using limactl snapshot makes it any better ?
@adonishong @cameronhudson8
